lib/media/segment_prefetch.js

```diff
@@ -0,0 +1,173 @@
+   *
+   * @param {(!shaka.media.SegmentReference)} startReference
+   * @public
+   */
+  prefetchSegments(startReference) {
+    if (!(this.prefetchLimit_ > 0)) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
+        this.stream_,
+    );
+    if (!this.stream_.segmentIndex) {
+      shaka.log.info(
+          logPrefix, 'missing segmentIndex',
+      );
+      return;
+    }
+    const currTime = startReference.startTime;
+    const maxTime = Math.max(currTime, this.prefetchPosTime_);
+    const iterator = this.stream_.segmentIndex.getIteratorForTime(maxTime);
+    let reference = startReference;
+    while (this.segmentPrefetchMap_.size < this.prefetchLimit_ &&
+            reference != null) {
+      if (!this.segmentPrefetchMap_.has(reference)) {
+        const op = this.fetchDispatcher_(reference, this.stream_);
+        this.segmentPrefetchMap_.set(reference, op);
+      }
+      this.prefetchPosTime_ = reference.startTime;
+      reference =iterator.next().value;
+    }
+  }
+
+  /**
+   * Get the result of prefetched segment if already exists.
+   * @param {(
+   *  !shaka.media.InitSegmentReference|!shaka.media.SegmentReference
+   * )} reference
+   * @return {?shaka.net.NetworkingEngine.PendingRequest} op
+   * @public
+   */
+  getPrefetchedSegment(reference) {
+    let op = null;
+    if (
+      !(reference instanceof shaka.media.SegmentReference) ||
+      !(this.prefetchLimit_ > 0)
+    ) {
+      return null;
+    }
+
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
+        this.stream_,
+    );
+
+    if (this.segmentPrefetchMap_.has(reference)) {
+      op = this.segmentPrefetchMap_.get(reference);
+      this.segmentPrefetchMap_.delete(reference);
+      shaka.log.info(
+          logPrefix,
+          'reused prefetched segment at time:', reference.startTime,
+          'mapSize', this.segmentPrefetchMap_.size);
+    } else {
+      shaka.log.info(
+          logPrefix,
+          'missed segment at time:', reference.startTime,
+          'mapSize', this.segmentPrefetchMap_.size);
+    }
+    return op;
+  }
+
+  /**
+   * Clear all segment data.
+   * @public
+   */
+  clearAll() {
+    if (this.segmentPrefetchMap_.size === 0) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
+        this.stream_,
+    );
+    for (const reference of this.segmentPrefetchMap_.keys()) {
+      const operation = this.segmentPrefetchMap_.get(reference);
+      this.segmentPrefetchMap_.delete(reference);
+      operation.abort();
+    }
+    shaka.log.info(logPrefix, 'cleared all');
+    this.prefetchPosTime_ = 0;
+  }
+
+  /**
+   * Reset the prefetchLimit and clear all internal states.
+   * Called by StreamingEngine when configure() was called.
+   * @param {number} prefetchLimit
+   * @public
+   */
+  resetLimit(prefetchLimit) {
+    if (prefetchLimit !== this.prefetchLimit_) {
+      this.clearAll();
```

Comment:
Design-wise, I'm not sure it makes sense to always clear all of the segments away in this case.
For example, if you are going from a pre-fetch limit of 4 to a pre-fetch limit of 6, couldn't you just keep the stuff you have already pre-fetched?

It seems like the most efficient thing to do would be to store an array that keeps track of the order of the pre-fetched segments, and then pop and clear segments from the end of it until you fit into the new limit. That way nothing happens if the limit has been increased, and you clear away as few pre-fetched segments as possible in the case of the limit being decreased.

Reply:
updated. actually it turns out that Map.keys reserves the insertion order.
so instead of creating a new ordered array, i just grab `segmentPrefetchMap_.keys` and iterate from end to pop + clear the segments.
not sure Array.from(iterator) will actually do any iteration though, but it might have minimal performance impact since resetLimit could be called very rarely.
open to other suggestions :)

Reply:
pushed a new commit to simplify resetLimit, by using while + arr.pop() instead of for loop from end of arr